src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseIntegrationTestService.java

```diff
@@ -2166,4 +2169,25 @@ void test_redirectGetTemplateRepositoryFilesWithContent(BiFunction<ProgrammingEx
     private long getMaxProgrammingExerciseId() {
         return programmingExerciseRepository.findAll().stream().mapToLong(ProgrammingExercise::getId).max().orElse(1L);
     }
+
```

Comment:
Would it make sense to extract this? Unsure if a default method on the MockDelegate would make sense.

But you use the very same method in the other TestService. The only difference is the way you retrieve the project key in the last two statements: Here, using the URL service, in the other TestService using the getter of the Exercise.

Reply:
time-wise, it is definitely not a good idea to invoke "findAll()". If this is the last test, that gets executed, there will be thousands of exercises which all are loaded into Java consuming lots of wasted time and way too much memory.
We should forbid the use of findAll in all places (production code and test code) except there is a really really good reason and no other way to achieve the desired functionality. Please change this here to make @DominikRemo and @laadvo happy as well ;-)

Reply:
- Replaced the findAll call with a more efficient solution
- Regarding the `setupMocksForConsistencyChecksOnImport`: I already looked into removing this code duplication, but don't see a good way to do it. Putting it as a default method in `MockDelegate` isn't a great solution IMO: It would be the only default method in there and would be highly specific. Every other method in `MockDelegate` is more general and tailored to mock a single action of the VCS/CI. Also, you would have to pass a service as a parameter, which I've not seen being done anywhere else in Artemis. I think the best solution would ultimately be to split up this file into multiple smaller test classes. It currently is massive (>2000 LOC) and not tailored to test the importing of programming exercises (which caused me some headaches). This would then also remove the code duplication, as you could simply put all import-related test cases in the same class.